chain-abci/src/app/rewards.rs

```diff
@@ -0,0 +1,193 @@
+use std::cmp::min;
+use std::f64::consts::E;
+
+use chain_core::init::{coin::Coin, params::NetworkParameters};
+use chain_core::state::account::StakedStateAddress;
+use chain_core::state::tendermint::TendermintValidatorAddress;
+use chain_core::tx::fee::Milli;
+
+use crate::app::{update_account, ChainNodeApp};
+use crate::enclave_bridge::EnclaveProxy;
+use crate::storage::tx::get_account;
+
+fn milli_to_f64(milli: Milli) -> f64 {
+    milli.as_millis() as f64 / 1000_f64
+}
+
+/// FIXME more stable arithmetic
+pub fn monetary_expansion(
+    tau: Milli,
+    total_staking: Coin,
+    minted: Coin,
+    params: &NetworkParameters,
+) -> Coin {
+    let cap = params.get_rewards_monetary_expansion_cap();
+    let r0 = milli_to_f64(params.get_rewards_monetary_expansion_r0());
+    let tau = milli_to_f64(tau);
+    let total_staking = u64::from(total_staking) as f64 / 1_0000_0000_f64;
+    let amount = total_staking * r0 * E.powf(-total_staking / tau);
+    min(
+        (cap - minted).unwrap_or_default(),
+        Coin::new(amount as u64).unwrap(),
+    )
+}
+
+type RewardsDistribution = Vec<(StakedStateAddress, Coin)>;
+impl<T: EnclaveProxy> ChainNodeApp<T> {
+    /// Distribute rewards pool
+    pub fn rewards_try_distribute(&mut self) -> Option<(RewardsDistribution, Coin)> {
+        let state = self.last_state.as_mut().unwrap();
+
+        if state.block_time < state.genesis_time
+            || state.block_time < state.rewards_pool.last_distribution_time
```

Comment:
Is the time in Tendermint guaranteed to be increasing? I lost the track of "BFT Time" stuff in Tendermint

Reply:
@devashishdxt

Reply:
Yeah, time monotonicity is guaranteed: https://docs.tendermint.com/master/spec/consensus/bft-time.html